krypto/src/commonMain/kotlin/com/soywiz/krypto/encoding/Base64.kt

```diff
@@ -1,42 +1,49 @@
 package com.soywiz.krypto.encoding
 
 object Base64 {
-    private val TABLE = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/="
-    private val DECODE = IntArray(0x100).apply {
-        for (n in 0..255) this[n] = -1
-        for (n in 0 until TABLE.length) {
-            this[TABLE[n].toInt()] = n
-        }
-    }
+    private const val TABLE = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/="
+    private const val TABLE_SAFE = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_="
 
     operator fun invoke(v: String) = decodeIgnoringSpaces(v)
     operator fun invoke(v: ByteArray) = encode(v)
 
-    fun decode(str: String): ByteArray {
+    fun decode(str: String, isUrlSafe: Boolean = false): ByteArray {
         val src = ByteArray(str.length) { str[it].toByte() }
         val dst = ByteArray(src.size)
-        return dst.copyOf(decode(src, dst))
+        return dst.copyOf(decode(src, dst, isUrlSafe))
     }
 
-    fun decodeIgnoringSpaces(str: String): ByteArray {
-        return decode(str.replace(" ", "").replace("\n", "").replace("\r", ""))
+    fun decodeIgnoringSpaces(str: String, isUrlSafe: Boolean = false): ByteArray {
+        return decode(str.replace(" ", "").replace("\n", "").replace("\r", ""), isUrlSafe)
     }
 
-    fun decode(src: ByteArray, dst: ByteArray): Int {
+    fun decode(src: ByteArray, dst: ByteArray, isUrlSafe: Boolean = false): Int {
+        val decoder = IntArray(0x100).apply {
+            val table = if (isUrlSafe) {
+                TABLE_SAFE
+            } else {
+                TABLE
+            }
+            for (n in 0..255) this[n] = -1
+            for (n in table.indices) {
+                this[table[n].toInt()] = n
+            }
+        }
+
```

Comment:
This data is immutable after setting it, so can be in an object. Better to extract this as part of the object, and do the if with the two precomputed DECODE tables.

Reply:
Done